### PR TITLE
Flatten nested errors

### PR DESCRIPTION
**Pull request by stevehodgkiss**
- Fixes an issue where by requesting errors for an attribute (with for) the Errors object would then think it had errors (with .any?)
- Flattens nested errors based on the parent attribute name.
- `#attribute` returns the nested attribute name, `#attribute_name` always returns the de-namespaced attribute name.

### CODE REVIEW

File: lib/lotus/validations/attribute.rb [hunk 1]
```diff
@@ -200,8 +200,10 @@ def size
       def nested
         _validate(__method__) do |validator|
           errors = value.validate
-          unless errors.empty?
-            @errors.set @name, errors
+          if errors.any?
```
Comment (AlfonsoUceda) [inline lib/lotus/validations/attribute.rb]: you can remove this if, if errors if empty the each loop won't be executed.

**Comment by AlfonsoUceda** [comment]
for now :+1:

File: lib/lotus/validations/attribute.rb [hunk 2]
```diff
@@ -200,8 +200,9 @@ def size
       def nested
         _validate(__method__) do |validator|
           errors = value.validate
-          unless errors.empty?
-            @errors.set @name, errors
+          errors.each do |error|
+            new_error = Error.new(error.attribute, error.validation, error.expected, error.actual, @name)
```
Comment (joneslee85) [inline lib/lotus/validations/attribute.rb]: 4 params is abit hard to follow, keyed params would be great for this constructor.

File: lib/lotus/validations/error.rb [hunk 3]
```diff
@@ -41,9 +58,10 @@ class Error
       #
       # @since 0.1.0
       # @api private
-      def initialize(attribute, validation, expected, actual)
-        @attribute, @validation, @expected, @actual =
-          attribute, validation, expected, actual
+      def initialize(attribute_name, validation, expected, actual, namespace = nil)
+        @attribute_name, @validation, @expected, @actual, @namespace =
```
Comment (joneslee85) [inline lib/lotus/validations/error.rb]: how about instance variables in many lines separately? According to benchmark by Pete Yandel, 1 line assignment is much slower

**Comment by joneslee85** [comment]
:+1:

File: lib/lotus/validations/error.rb (hunk 3 above)
Comment (stevehodgkiss) [inline lib/lotus/validations/error.rb]: Yeah I would prefer these to be on one line too. Will update...

**Comment by joneslee85** [comment]
Looking great!

:+1:  from me

Waitin' for @jodosha 's review

**Comment by jodosha** [comment]
I think this PR deserves a public recap of the discussion happened between me and @stevehodgkiss 

The goals are:

  * Keep bulk operations around `validator.errors.each`. It yields a flatten list of errors.
  * Nested attributes can cause name clashing (see Example 1), `Error#attribute` will return the complete name of the attribute (Example 2).
  * The "dot notation" described above is unhandy if we want only to get the attribute name. For this reason we have introduced `#attribute_name` (Example 3).

```ruby
class Signup
  include Lotus::Validations

  attribute :name, presence: true
  attribute :programming_language do
    attribute :name, presence: true
  end
end

validator = Signup.new({})
validator.valid?

# Example 1
validator.errors.each do |error|
  error.attribute # => Without this PR, it always returned 'name'. This makes hard to understand where it belongs to (top level or 'programming_language').
end

# Example 2
validator.errors.each do |error|
  error.attribute # => first iteration: "name", second iteration: "programming_language.name".
end

# Example 3
validator.errors.each do |error|
  error.attribute_name # => first iteration: "name", second iteration: "name".
end
```

We decided to add a fifth argument to `Error#initialize`, in this way, `Error` is responsible to understand and provide simple and complex names via the "dot notation".

**Comment by jodosha** [comment]
The other notable change is the notation to fetch nested errors:

```ruby
# OLD
validator.errors.for(:address).for(:line_one)

# NEW
validator.errors.for('address.line_one')
```

**Comment by jodosha** [comment]
This looks good :+1: Thanks @stevehodgkiss Merging.
